### PR TITLE
Initial version of anisble extension to run play book from either ans…

### PR DESCRIPTION
**Pull request by hililh**
Initial version of anisble extension to run play book from either ansible cli or tower
TODO: 
1. Fixing node and gulpfile
2. displaying dropdown instead of text field for job template name

### CODE REVIEW

File: Extensions/Ansible/Src/Tasks/Ansible/main.ts [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+
+const  btoa  =  require('abab').btoa;
+
+var httpClient = require('vso-node-api/HttpClient');
+var httpObj = new httpClient.HttpCallbackClient(tl.getVariable("AZURE_HTTP_USER_AGENT"));
+
+try {
+    tl.setResourcePath(path.join(__dirname, "task.json"));
+} catch (err) {
+    tl.setResult(tl.TaskResult.Failed, tl.loc("TaskNotFound", err));
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/main.ts]: No need to do process exit etc

File: Extensions/Ansible/Src/Tasks/Ansible/main.ts [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+            jobId = response.body['id'];
+        } else {
+            throw (tl.loc('CouldnotLaunchJob', response.statusCode, response.statusMessage));
+        }
+        return jobId;
+    }
+
+    public async execute() {
+        try {
+            this._jobTemplateId = await this.GetJobTemplateId();
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/main.ts]: have the function name start with lowercase

File: Extensions/Ansible/Src/Tasks/Ansible/main.ts [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+            this.LaunchJob().then(async (jobId) => {
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/main.ts]: YOu should say await this.LaunchJob();

File: Extensions/Ansible/Src/Tasks/Ansible/main.ts [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+                tl.setResult(tl.TaskResult.Failed, error);
+            });
+        }
+        catch (error) {
+            tl.setResult(tl.TaskResult.Failed, error);
+        }
+    }
+
+
+    public async beginRequest(request: WebRequest): Promise<WebResponse> {
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/main.ts]: they should be private

File: Extensions/Ansible/Src/Tasks/Ansible/main.ts [hunk 5]
```diff
@@ -0,0 +1,456 @@
+/// <reference path="../../../../../definitions/node.d.ts" />
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/main.ts]: UTs

**Review by sachinma — CHANGES_REQUESTED** [review]

File: Extensions/Ansible/Src/Tasks/Ansible/ssh2helpers.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    var cmdToRun = command;
+    if(cmdToRun.indexOf(';') > 0) {
+        //multiple commands were passed separated by ;
+        cmdToRun = cmdToRun.replace(/;/g, '\n');
+    }
+    tl.debug('cmdToRun = ' + cmdToRun);
+
+    sshClient.exec(cmdToRun, (err, stream) => {
+        if(err) {
+            defer.reject(tl.loc('RemoteCmdExecutionErr', err))
```
Comment (amaljg) [inline Extensions/Ansible/Src/Tasks/Ansible/ssh2helpers.ts]: Shouldn't we return here?

File: Extensions/Ansible/Src/vss-extension.json [hunk 7]
```diff
@@ -0,0 +1,80 @@
+{
+  "manifestVersion": 1.0,
+  "id": "vss-services-ansible",
+  "name": "Ansible",
+  "publisher": "Microsoft",
```
Comment (amaljg) [inline Extensions/Ansible/Src/vss-extension.json]: Is this the correct publisher?

File: Extensions/Ansible/Src/Tasks/Ansible/package.json [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+  "name": "ansible",
+  "version": "0.0.1",
+  "description": "Ansible task for playbook execution",
+  "main": "main.js",
+  "author": "Microsoft",
+  "license": "ISC",
+  "dependencies": {
+    "scp2": "^0.5.0",
+    "ssh2": "^0.5.5",
+    "abab": "^1.0.3",
```
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/package.json]: Done

File: Extensions/Ansible/Src/Tasks/Ansible/package.json [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    "vso-node-api": "6.0.1-preview",
+    "vsts-task-lib" : "^0.9.20"
```
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/package.json]: Done

File: Extensions/Ansible/Src/vss-extension.json [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+      "_description.comment": "Don't Localize the word 'Ansible Tower'",
+      "type": "ms.vss-endpoint.service-endpoint-type",
+      "targets": [ "ms.vss-endpoint.endpoint-types" ],
+      "properties": {
+        "name": "ansibletower",
+        "displayName": "Ansible Tower",
+        "endpointTypeId": "AnsibleTower",
+        "authenticationSchemes": [
+          {
+            "type": "ms.vss-endpoint.endpoint-auth-scheme-basic"
```
Comment (hililh) [inline Extensions/Ansible/Src/vss-extension.json]: Only basic is supported

File: Extensions/Ansible/Src/Tasks/Ansible/main.ts (hunk 5 above)
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/main.ts]: Will do in another PR

File: Extensions/Ansible/Src/Tasks/Ansible/main.ts [hunk 11]
```diff
@@ -0,0 +1,456 @@
...
+/// <reference path="../../../../../definitions/Q.d.ts" />
+
+import tl = require("vsts-task-lib/task");
+import path = require("path");
+import querystring = require('querystring');
+import Q = require("q");
+import util = require("util");
+
+import sshHelper = require('./ssh2helpers');
+import { RemoteCommandOptions } from './ssh2helpers'
```
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/main.ts]: Removed

File: Extensions/Ansible/Src/Tasks/Ansible/main.ts (hunk 1 above)
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/main.ts]: Removed

File: Extensions/Ansible/Src/Tasks/Ansible/main.ts [hunk 12]
```diff
@@ -0,0 +1,456 @@
...
+}
+
+class WebResponse {
+    public statusCode;
+    public headers;
+    public body;
+    public statusMessage;
+}
+
+class AnsibleTowerConstants {
```
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/main.ts]: Done

File: Extensions/Ansible/Src/Tasks/Ansible/main.ts [hunk 13]
```diff
@@ -0,0 +1,456 @@
...
+
+class AnsibleInterface {
+    constructor() {
+    }
+    public execute(): void {
+        throw "selected interface not supported";
+    }
+}
+
+class AnsibleCommandLineInterface extends AnsibleInterface {
```
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/main.ts]: Done

File: Extensions/Ansible/Src/Tasks/Ansible/main.ts (hunk 2 above)
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/main.ts]: Done

File: Extensions/Ansible/Src/Tasks/Ansible/main.ts (hunk 3 above)
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/main.ts]: Done

File: Extensions/Ansible/Src/Tasks/Ansible/main.ts (hunk 4 above)
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/main.ts]: Done

File: Extensions/Ansible/Src/Tasks/Ansible/package.json [hunk 14]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "ansible",
+  "version": "0.0.1",
+  "description": "Ansible task for playbook execution",
```
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/package.json]: It is just the package.json file for npm and node dependencies

File: Extensions/Ansible/Src/Tasks/Ansible/main.ts [hunk 15]
```diff
@@ -1,456 +1,35 @@
...
+        var ansibleInterface: AnsibleInterface = AnsibleInterfaceFactory.GetAnsibleInterface(tl.getInput("ansibleInterface", true));
+        if (ansibleInterface) {
+            ansibleInterface.execute();
+        } else {
+            tl.setResult(tl.TaskResult.Failed, "");
+        }
+    } catch (error) {
+        tl.setResult(tl.TaskResult.Failed, error);
+    }
+}
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/main.ts]: move setresource to this file

File: Extensions/Ansible/Src/Tasks/Ansible/AnsibleInterface.ts [hunk 16]
```diff
@@ -0,0 +1,15 @@
+var os = require('os');
+
+export class AnsibleInterface {
+    constructor() {
+    }
+    public execute(): void {
+        throw "selected interface not supported";
+    }
+    protected _writeLine(str) : void {
+        this._outStream.write(str + os.EOL);
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/AnsibleInterface.ts]: This should not be required

File: Extensions/Ansible/Src/Tasks/Ansible/AnsibleInterface.ts [hunk 17]
```diff
@@ -0,0 +1,15 @@
+var os = require('os');
+
+export class AnsibleInterface {
+    constructor() {
+    }
+    public execute(): void {
+        throw "selected interface not supported";
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/AnsibleInterface.ts]: consider returning Promise<void>

File: Extensions/Ansible/Src/Tasks/Ansible/AnsibleCommandLineInterface.ts [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+        var cmdToRun = command;
+        if (cmdToRun.indexOf(';') > 0) {
+            //multiple commands were passed separated by ;
+            cmdToRun = cmdToRun.replace(/;/g, '\n');
+        }
+        tl.debug('cmdToRun = ' + cmdToRun);
+
+        sshClient.exec(cmdToRun, (err, stream) => {
+            if (err) {
+                defer.reject(tl.loc('RemoteCmdExecutionErr', err))
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/AnsibleCommandLineInterface.ts]: There should be a return statement or the next clause should be in else

File: Extensions/Ansible/Src/Tasks/Ansible/AnsibleTowerInterface.ts [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+        return ansibleJobEventUrl;
+    }
+
+    private async getJobTemplateId(): Promise<string> {
+
+        var jobTemplateId: string = null;
+        var request = new WebRequest();
+        request.method = 'GET';
+        request.uri = util.format(this._jobTemplateIdUrlFormat, this._hostname, this._jobTemplateName);
+        request.body = this.getRequestData();
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/AnsibleTowerInterface.ts]: getEmptyRequestData

Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/AnsibleTowerInterface.ts]: body should not be required for GET

File: Extensions/Ansible/Src/Tasks/Ansible/AnsibleTowerInterface.ts [hunk 20]
```diff
@@ -0,0 +1,269 @@
...
+import util = require("util");
+
+import { AnsibleInterface } from './AnsibleInterface';
+
+var httpClient = require('vso-node-api/HttpClient');
+var httpObj = new httpClient.HttpCallbackClient(tl.getVariable("AZURE_HTTP_USER_AGENT"));
+
+tl.setResourcePath(path.join(__dirname, "task.json"));
+
+class WebRequest {
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/AnsibleTowerInterface.ts]: No type information here

File: Extensions/Ansible/Src/Tasks/Ansible/AnsibleTowerInterface.ts [hunk 21]
```diff
@@ -0,0 +1,269 @@
...
+            }
+            if (this.isJobInTerminalState(status)) {
+                break;
+            }
+
+            await this.sleepFor(waitTimeInSeconds);
+
+            timeElapsed += waitTimeInSeconds;
+            if (timeElapsed >= longRunningJobThreshold) {
+                waitTimeInSeconds = 120;
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/AnsibleTowerInterface.ts]: Lets not exceed by 1 minute

File: Extensions/Ansible/Src/Tasks/Ansible/AnsibleTowerInterface.ts [hunk 22]
```diff
@@ -0,0 +1,269 @@
...
+        var response = await this.beginRequest(request);
+        if (response.statusCode === 200) {
+            status = response.body['status'];
+        } else {
+            throw (tl.loc('FailedToGetJobDetails', response.statusCode, response.statusMessage));
+        }
+        return status;
+    }
+
+    private async getJobEvent(jobId: string): Promise<string[]> {
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/AnsibleTowerInterface.ts]: rename to events. and take another input for startEventId

**Review by sachinma — CHANGES_REQUESTED** [review]

File: Extensions/Ansible/Src/Tasks/Ansible/main.ts (hunk 15 above)
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/main.ts]: Done

File: Extensions/Ansible/Src/Tasks/Ansible/AnsibleCommandLineInterface.ts (hunk 18 above)
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/AnsibleCommandLineInterface.ts]: Done

File: Extensions/Ansible/Src/Tasks/Ansible/main.ts (hunk 15 above)
Comment (hililh) [inline Extensions/Ansible/Src/Tasks/Ansible/main.ts]: Done

File: Extensions/Ansible/Src/Tasks/Ansible/AnsibleTowerInterface.ts [hunk 23]
```diff
@@ -0,0 +1,266 @@
...
+        while (jobEventUrl) {
+            request.uri = jobEventUrl;
+            var response = await this.beginRequest(request);
+            if (response.statusCode === 200) {
+                var totalEvents = response.body['count'];
+                var results: any[] = response.body['results'];
+                var nextPageUrl = response.body['next'];
+                results.forEach((event) => {
+                    if (event['counter'] > lastDisplayedEvent)
+                        stdoutArray[event['counter']] = event['stdout'];
```
Comment (sachinma) [inline Extensions/Ansible/Src/Tasks/Ansible/AnsibleTowerInterface.ts]: Add a comment explaining why do you need to do it.

**Review by sachinma — APPROVED** [review]
